### DISCUSSION TITLE
Configuring comments with DISQUS

### DISCUSSION THREAD

**Discussion by mmarroquinavatar** [requarks/wiki#3454] [Help / Questions] [answered]
**Question**
Hi, first at all, thanks for so wonderful product. This wiki software is really easy to use and accessible for technical and non-technical users.

Currently I have deployed Wiki.js and my unique problem so far is the comments section. The default comments engine obviously lacks feature but we have a nice alternative like DISQUS. Now when I enabled DISQUS and I published a message in a page, I see that the same thread appears in all the pages. I know this issue happens normally with single page applications. Can anyone provide a tip about how to customize Wiki.js to display inidividual threads? 

Thanks for your help.

**Host Info (please complete the following information):**
OS: Docker
Wiki.js version: 2.5.159
Database engine: postgres

**Comment by mmarroquinavatar** [comment]
Hi, after some search I found a Wiki.js site with a custom implementation of DISQUS. After some analysis I found that they inject this variables to the DISQUS script instead of the current static variables used in the current release. It works wonderful, now each article have a different DISQUS forum.

    var disqus_config = function () {
      this.page.url = WIKI.$el.baseURI;
      this.page.identifier = WIKI.$store.get('page/id');
    };
 
Do you think is it a good idea to implement DISQUS this way?

**Reply by EyalSeagull** [reply]
Hi @mmarroquinavatar ,

Your solution worked like a charme. I just had to update the file 'wiki/server/modules/comments/disqus/code.yml' to this:

```
main: |
  <div id="disqus_thread"></div>
body: |
  <script>
    var disqus_config = function () {
      this.page.url = WIKI.$el.baseURI;
      this.page.identifier = WIKI.$store.get('page/id');
    };  
    (function() {
      var d = document, s = d.createElement('script');
      s.src = 'https://{{accountName}}.disqus.com/embed.js';
      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    })();
  </script>
```


Thank you very much!!

**Reply by EyalSeagull** [reply]
@NGPixel , it seems like a fix to DISQUS problem. Is there another solution?

**Reply by NGPixel** [reply]
While that code may works in your browser, a race condition is very likely and it's possible for the WIKI instance to not exist yet when that code is executed.

The page ID is exactly the same so that's not the issue. The only difference with the original code is the page URL which outputs the `/i/123` shorthand version instead of the full path, which seems to cause issues with Disqus.

A proper solution would be:
```yaml
main: |
  <div id="disqus_thread"></div>
body: |
  <script>
    var disqus_config = function () {
      this.page.url = window.location.href;
      this.page.identifier = '{{pageId}}';
    };  
    (function() {
      var d = document, s = d.createElement('script');
      s.src = 'https://{{accountName}}.disqus.com/embed.js';
      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    })();
  </script>
```

If you can confirm the above code works fine for you, either submit a PR or I can patch it directly.

**Reply by EyalSeagull** [reply]
@NGPixel , `code.yml` was altered as specified. Problem continue. It seems like the `{{pageId}}` is always the same, no matter what page we're in.

For exemple: opening 3 different pages resulted in the same JS code, with the **same** page id (in this case '39'):
```
var disqus_config = function () {
    this.page.url = window.location.href;
    this.page.identifier = '39';
  };  
  (function() {
    var d = document, s = d.createElement('script');
    s.src = 'https://kesem.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
```

**Server:**
OS Docker Container (Linux)
Wiki.js 2.5.277
Node.js 16.14.0
PostgreSQL 11.15
RAM 966.66 MB

**I've tried with the following clients:**
Chrome: 100.0.4896.75 (Official) 64 bits
Firefox: 99.0 64 bits
Microsoft Edge: 100.0.1185.36 (Official) (64 bits)

**Reply by EyalSeagull** [reply]
Hi @NGPixel , any idea of why the {{pageId}} is the same in all the pages? Thanks.

**Reply by NGPixel** [reply]
I could reproduce the ID being always the same, which is a bug.

Just found out that the code which injects the variables is completely wrong and would always set itself to the first page being loaded. I'll push an update soon with the fix.

**Comment by NGPixel** [comment, marked as answer]
Fixed by a647626a51da8aec5adf35123af5f1dce656fa48. Will be included in the next release.

**Reply by EyalSeagull** [reply]
@NGPixel , `{{pageId}}` shows the correct information now. Thank you very much.
